aten/src/ATen/FunctionalStorageImpl.cpp

```diff
@@ -90,10 +91,23 @@ bool Alias::apply_updates() {
   return any_updates;
 }
 
+c10::SymInt get_nbytes(const Tensor& value) {
+  if (value.unsafeGetTensorImpl()->has_symbolic_sizes_strides()) {
+    // Today, the two implementations of SymInt are in Python (proxy tensor),
+    // and lazy tensor (LTC/XLA).
+    // LTC hasn't implemented SymInt support yet though (torch::lazy::SymIntNodeImpl).
+    // Once it does, we should remove this check.
+    if (value.key_set().has(c10::DispatchKey::Python)) {
+      return at::detail::computeStorageNbytes(value.sym_sizes(), value.sym_strides(), value.dtype().itemsize(), value.storage_offset());
+    }
+  }
+  return at::detail::computeStorageNbytes(value.sizes(), value.strides(), value.dtype().itemsize(), value.storage_offset());
+}
```

Comment:
cc @bdhirsh 

I think this code is a bit questionable. In particular, in https://github.com/pytorch/pytorch/pull/84246/commits/146c74b26a43f4107d020b0d2d4f97cebd2ac1e2 I replaced this with directly reading out the nbytes the storage on value. This avoids doing a bunch of unnecessary size manipulations in the IR.

I'd like to know if this is generally correct. But to know this, we need to know what the meaning of `FunctionalStorageImpl(const Tensor& value)`. What exactly is this storage denoting? IIUC, it is called when we a turning a tensor into a functional wrapper tensor, e.g., at

```
FunctionalTensorWrapper::FunctionalTensorWrapper(const Tensor& value)
  : c10::TensorImpl(
      c10::Storage(c10::make_intrusive<functionalization::FunctionalStorageImpl>(value)),
      c10::DispatchKeySet(DispatchKey::Functionalize) | value.key_set(),
      value.dtype()
    ),
    value_(value)
{
  set_constructor_metadata();
}
```

In that case, reading out the nbytes is right. So the question is primarily if there are other cases we exercise this.

(BTW the linked commit doesn't go far enough, we should probably read nbytes directly for the non dynamic case too)

Reply:
oh yep - using `.nbytes()` and `.sym_nbytes()` on the inner tensor's storage feels way better.

Although I don't think we'll be able to do this for the LTC/XLA case yet, because IIRC XLA and LTC both have null storages. Maybe we should just make them more similar to meta tensor storages, and give them a valid storage that tracks nbytes but with a null data pointer.

Reply:
But XLA/LTC should also not have to "wrap" a tensor in a functional tensor wrapper, and you can compute default contiguous storage size in that case

Reply:
oh - you're saying that for the LTC/XLA case, we're still forced to compute the storage size here, but it's better to just assume contiguous storages (so we get a cleaner IR)?

Reply:
Well, what actually should be happening is that you're creating a tensor from a factory call like empty() or empty_strided(), and those operators should themselves be responsible for computing storage size.